src/Interpreters/AsynchronousMetrics.cpp

```diff
@@ -63,15 +63,45 @@ void AsynchronousMetrics::run()
 {
     setThreadName("AsyncMetrics");
 
-    /// Next minute + 30 seconds. To be distant with moment of transmission of metrics, see MetricsTransmitter.
-    const auto get_next_minute = []
+    const auto get_next_update_time = []
     {
-        return std::chrono::time_point_cast<std::chrono::minutes, std::chrono::system_clock>(
-            std::chrono::system_clock::now() + std::chrono::minutes(1)) + std::chrono::seconds(30);
+        using namespace std::chrono;
+
+        // The period doesn't really have to be configurable, but sometimes you
+        // need to change it by recompilation to debug something. The generic
+        // code is left here so that you don't have to ruin your mood by touching
+        // std::chrono.
+        const seconds period(60);
+
+        const auto now = time_point_cast<seconds>(system_clock::now());
+
+        // Use seconds since the start of the hour, because we don't know when
+        // the epoch started, maybe on some weird fractional time.
+        const auto start_of_hour = time_point_cast<seconds>(time_point_cast<hours>(now));
+        const auto seconds_passed = now - start_of_hour;
+
+        // Rotate time forward by half a period -- e.g. if a period is a minute,
+        // we'll collect metrics on start of minute + 30 seconds. This is to
+        // achieve temporal separation with MetricTransmitter. Don't forget to
+        // rotate it back.
+        const auto rotation = period / 2;
+
+        const auto periods_passed = (seconds_passed + rotation) / period;
+        const auto seconds_next = (periods_passed + 1) * period - rotation;
+        const auto time_next = start_of_hour + seconds_next;
+
+        return time_next;
```

Comment:
Funny thing, but previous two rows were much easier to read even without comments. Probably, it's just because code is more general now.